### DISCUSSION TITLE
Can't import module: Get-MonkeySupportedService and New-O365Object not recognized

### DISCUSSION THREAD

**Discussion by nickchristie** [silverhack/monkey365#66] [Q&A] [answered]
Hi, sorry if I have missed something obvious, but I can't get Monkey365 to import. I am not a regular user of PowerShell.

`PS C:\Program Files\WindowsPowerShell\Modules> Import-Module monkey365`

`Get-MonkeySupportedService: The term 'Get-MonkeySupportedService' is not recognized as a name of a cmdlet, function, script file, or executable program. Check the spelling of the name, or if a path was included, verify that the path is correct and try again.
Get-MonkeySupportedService: The term 'Get-MonkeySupportedService' is not recognized as a name of a cmdlet, function, script file, or executable program. Check the spelling of the name, or if a path was included, verify that the path is correct and try again.
New-O365Object: The term 'New-O365Object' is not recognized as a name of a cmdlet, function, script file, or executable program. Check the spelling of the name, or if a path was included, verify that the path is correct and try again.`

I have tried it on a 'vanilla' Windows Server 2022 VM in Azure, with a fresh install of PowerShell 7.3.6, as well as on my Windows 10 PC.
I unzipped https://github.com/silverhack/monkey365/archive/refs/heads/main.zip to C:\Program Files\WindowsPowerShell\Modules\monkey365, ran Unblock-File as instructed, and Set-ExecutionPolicy to Unrestricted for good measure. I am an administrator on both systems.
The file "C:\Program Files\WindowsPowerShell\Modules\monkey365\core\utils\Get-MonkeySupportedService.ps1" does exist, but I can't find any file under monkey365 called "New-O365Object".

If I try to Invoke-Monkey365 anyway, I get loads of errors, starting with
`InvalidOperation: C:\Program Files\WindowsPowerShell\Modules\monkey365\Invoke-Monkey365.ps1:305
Line |
 305 |              Azure = $Script:azure_plugins
     |                      ~~~~~~~~~~~~~~~~~~~~~
     | The variable '$Script:azure_plugins' cannot be retrieved because it has not been set.`

$env:PSModulePath is set to:
`C:\Users\FCDeveloper\Documents\PowerShell\Modules;C:\Program Files\PowerShell\Modules;c:\program files\powershell\7\Modules;C:\Program Files\WindowsPowerShell\Modules;C:\Windows\system32\WindowsPowerShell\v1.0\Modules;C:\Program Files (x86)\Microsoft SQL Server\150\Tools\PowerShell\Modules\`

Thanks for any help you can give!

**Comment by silverhack** [comment]
Hi @nickchristie,  

As far as I can see probably when you unzip the main folder is unzipped into a "Monkey365-main" folder. You must ensure that the folder's name is Monkey365. Detailed steps about how to install the module is located here [https://silverhack.github.io/monkey365/install-instructions/](https://silverhack.github.io/monkey365/install-instructions/)

**Reply by nickchristie** [reply]
Thanks for your response, but unfortunately no that isn't the reason for the errors. I noticed that I had to rename the folder after unzipping:

```
PS C:\Program Files\WindowsPowerShell\Modules\monkey365> ls

    Directory: C:\Program Files\WindowsPowerShell\Modules\monkey365

Mode                 LastWriteTime         Length Name
----                 -------------         ------ ----
d----          16/08/2023    18:11                .github
d----          16/08/2023    18:11                config
d----          16/08/2023    18:11                core
d----          16/08/2023    18:11                docker
d----          16/08/2023    18:11                docs
d----          16/08/2023    18:11                images
d----          16/08/2023    18:11                plugins
d----          16/08/2023    18:11                rules
d----          16/08/2023    18:11                tests
-a---          16/08/2023    18:11             96 .dockerignore
-a---          16/08/2023    18:11            364 .gitignore
-a---          16/08/2023    18:11           4841 action.yml
-a---          16/08/2023    18:11       15259790 assets.zip
-a---          16/08/2023    18:11           2492 build.ps1
-a---          16/08/2023    18:11          24937 Invoke-Monkey365.ps1
-a---          16/08/2023    18:11          11357 LICENSE
-a---          16/08/2023    18:11           3646 mkdocs.yml
-a---          16/08/2023    18:11            702 monkey365.ps1
-a---          16/08/2023    18:11           5089 monkey365.psd1
-a---          16/08/2023    18:11           1573 monkey365.psm1
-a---          16/08/2023    18:11           5112 README.md
-a---          16/08/2023    18:11             81 requirements-docs.txt
```
I was wrong about the absence of New-O365Object: it is in fact present as "C:\Program Files\WindowsPowerShell\Modules\monkey365\core\init\New-O365Object.ps1".

The pattern of error messages suggests that the problems are arising on lines 31, 33 and 38 of "...\monkey365\monkey365.psm1" but that's just a guess.

**Reply by silverhack** [reply]
Hey @nickchristie 

Thanks for confirming! I'll try to install a fresh 2k12 server and I will let you know once I test the module. 

Thanks

**Reply by nickchristie** [reply]
I'm getting the same errors on my Windows 10 Pro PC as well.
I tried the Windows 2012 Server to check that it wasn't some strange behaviour just on my PC.
Both machines have the latest stable PowerShell release 7.3.6.

**Reply by silverhack** [reply]
Hi @nickchristie that's strange.... I have different environments and in all of them the module is working as expected (Win10, Win11, Ubuntu WSL and Ubuntu). I'll try to install the module in the path you suggest in order to understand it. 

Thanks for the information!

**Reply by silverhack** [reply]
Hey again @nickchristie, 

That is a very good first issue xP. It seems that when the module is placed in %ProgramFiles%\WindowsPowerShell\Modules there are some internal methods that are not available for an object. On the other hand, the internal search criteria fails, because the monkey365 place all internal modules in a folder called **modules**, and it breaks all things when the module is located in a path with a folder called "Modules".

So fast solution would be to move monkey365 module to your **user profile** (%HOMEDRIVE%%HOMEPATH%\Documents\) or another directory within your profile which not include a folder called **modules**. 

I'll add a new issue for this. 

Thanks!
Cheers

**Reply by nickchristie** [reply]
I didn't mention it but I also tried the other directory that PowerShell modules are commonly stored in: %HOMEPATH%\Documents\PowerShell\Modules but of course this didn't work either, for the same reason you discovered.
I now have it in "C:\Program Files\monkey365" from which I can import it just fine, thank you.

**Reply by silverhack** [reply]
Thanks a lot @nickchristie for discovering this, it's really appreciated. 

I'll patch the dev branch as soon as I can. 

Cheers

**Comment by silverhack** [comment, marked as answer]
Hi @nickchristie 

A new issue was created [here](https://github.com/silverhack/monkey365/issues/67)

Thanks!
